components/server/src/api/server.ts

```diff
@@ -174,9 +180,25 @@ export class API {
 
                     const context = args[1] as HandlerContext;
 
+                    const rateLimit = async (userId: string) => {
+                        try {
+                            const key = `${grpc_service}.${grpc_method}`;
+                            const options = self.config.rateLimits[key] || RateLimited.getOptions(target, prop);
+                            await self.getRateLimitter(options).consume(`${userId}_${key}`);
+                        } catch (e) {
+                            throw new ConnectError("rate limit exceeded", Code.ResourceExhausted, {
+                                // http compatibility, can be respected by gRPC clients as well
+                                // instead of doing an ad-hoc retry, the client can wait for the given amount of seconds
+                                "Retry-After": e.msBeforeNext / 1000,
+                            });
+                        }
+                    };
+
                     const apply = async <T>(): Promise<T> => {
-                        const user = await self.verify(context);
-                        context.user = user;
+                        const userId = await self.verify(context);
+                        context.userId = userId;
```

Comment:
@geropl I removed fetching a user here, we only verify JWT token and then use its subject. Services call pull as necessary? Do you see any security implications with it?

Reply:
For the FGA migration-thing, my understanding is that we still need to call [this code](https://github.com/gitpod-io/gitpod/blob/0fd84a51641bf64ef7ce08d3be9d8f7af8c9665a/components/server/src/user/user-service.ts#L89-L94), which we so far did by calling `userService.findUserById`.

For the SubjectId-topic: I'd try to remove all references to `userId` at all, as it blocks us in coming up with truly subject-agnostic APIs (as in: it does not whether if you are a admin, a user, an org or just a token: You can use a single API method, and get the same results, as long as you have access).
I will make a small suggestion PR now to illustrrate what I mean based on latest `main`.

Reply:
It would be nice if we can reconsider migration and avoid unnecessary fetching a user, but I will put it back.

Reply:
>For the SubjectId-topic

It does not change rate limiting right? We will always rate limit by subject id.

Reply:
> We will always rate limit by subject id.

:100: Absolutely! Just wanted to highlight it might be a userId or not. :+1:

Reply:
@svenefftinge @geropl Actually SessionHandler.verify does not ensure FGA migration. It does not call UserService.getLoggedInUser but DB object directly. We should use then UserService?

Reply:
ignore, I'm wrong